src/Compilers/CSharp/Test/Emit2/Semantics/LockTests.cs

```diff
@@ -0,0 +1,1797 @@
+                    {
+                        public void Dispose() { }
+                    }
+                }
+            }
+            """;
+        CreateCompilation(source).VerifyDiagnostics(
+            // (2,7): error CS0656: Missing compiler required member 'System.Threading.Lock.EnterLockScope'
+            // lock (l) { }
+            Diagnostic(ErrorCode.ERR_MissingPredefinedMember, "l").WithArguments("System.Threading.Lock", "EnterLockScope").WithLocation(2, 7));
+    }
+
+    [Fact]
+    public void GenericScope()
+    {
+        var source = """
+            System.Threading.Lock l = new();
+            lock (l) { }
+
+            namespace System.Threading
+            {
+                public class Lock
+                {
+                    public Scope<int> EnterLockScope() => new Scope<int>();
+
+                    public ref struct Scope<T>
+                    {
+                        public void Dispose() { }
+                    }
+                }
+            }
+            """;
+        CreateCompilation(source).VerifyDiagnostics(
+            // (2,7): error CS0656: Missing compiler required member 'System.Threading.Lock.EnterLockScope'
+            // lock (l) { }
+            Diagnostic(ErrorCode.ERR_MissingPredefinedMember, "l").WithArguments("System.Threading.Lock", "EnterLockScope").WithLocation(2, 7));
+    }
+
+    [Fact]
+    public void LockStruct()
+    {
+        var source = """
+            System.Threading.Lock l = new();
+            lock (l) { }
+
+            namespace System.Threading
+            {
+                public struct Lock
+                {
+                    public Scope EnterLockScope() => new Scope();
+
+                    public ref struct Scope
+                    {
+                        public void Dispose() { }
+                    }
+                }
+            }
+            """;
+        CreateCompilation(source).VerifyDiagnostics(
+            // (2,7): error CS0185: 'Lock' is not a reference type as required by the lock statement
+            // lock (l) { }
+            Diagnostic(ErrorCode.ERR_LockNeedsReference, "l").WithArguments("System.Threading.Lock").WithLocation(2, 7));
+    }
+
+    [Theory, CombinatorialData]
+    public void ScopeRegularStruct(bool implementsIDisposable)
+    {
+        var source = $$"""
+            static class Program
+            {
+                static void Main()
+                {
+                    System.Threading.Lock l = new();
+                    lock (l) { System.Console.Write("L"); }
+                }
+            }
+
+            namespace System.Threading
+            {
+                public class Lock
+                {
+                    public Scope EnterLockScope()
+                    {
+                        Console.Write("E");
+                        return new Scope();
+                    }
+
+                    public struct Scope {{(implementsIDisposable ? ": IDisposable" : "")}}
```

Comment:
>public struct Scope {{(implementsIDisposable ? ": IDisposable" : "")}} [](http://example.com/codeflow?start=20&length=70)

Should we request the `Scope` to be a ref struct?

Reply:
Not sure. I chose to not request that since the feature doesn't need it. But thinking about it more, if `Scope` wasn't a ref struct, we would want to forbid `await`s in `using`s on `Lock.Scope`, so the spec/feature would need changing anyway.